### PR TITLE
Added dialog to let user delete all failed messages

### PR DESCRIPTION
**Pull request by CyanBlob**
Long pressing a conversation now has an extra menu item to let the user delete failed messages from all conversations.

### CODE REVIEW

File: QKSMS/build.gradle [hunk 1]
```diff
@@ -3,7 +3,7 @@ buildscript {
         mavenCentral()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:1.3.0'
+        classpath 'com.android.tools.build:gradle:1.2.3'
```
Comment (moezbhatti) [inline QKSMS/build.gradle]: You should update your gradle

File: QKSMS/src/main/java/com/moez/QKSMS/common/DialogHelper.java [hunk 2]
```diff
@@ -21,6 +22,11 @@ public static void showDeleteConversationDialog(MainActivity context, long threa
         threadIds.add(threadId);
         showDeleteConversationDialog(context, threadIds);
     }
+    public static void showDeleteFailedConversationDialog(MainActivity context, long threadId) {
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/common/DialogHelper.java]: No space between this and the previous method

Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/common/DialogHelper.java]: `showDeleteFailedMessagesDialog` probably makes more sense

File: QKSMS/src/main/java/com/moez/QKSMS/common/DialogHelper.java [hunk 3]
```diff
@@ -70,4 +76,23 @@ public void onClick(View v) {
                 }).setNegativeButton(R.string.cancel, null)
                 .show(context.getFragmentManager(), QKDialog.CONFIRMATION_TAG);
     }
+
+    public static void showDeleteFailedConversationDialog(final MainActivity context, final List<Long> threadIds) {
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/common/DialogHelper.java]: `showDeleteFailedMessagesDialog` again

File: QKSMS/src/main/java/com/moez/QKSMS/data/Conversation.java [hunk 4]
```diff
@@ -32,6 +32,7 @@
 import com.moez.QKSMS.common.utils.PhoneNumberUtils;
 import com.moez.QKSMS.transaction.NotificationManager;
 import com.moez.QKSMS.transaction.SmsHelper;
+import com.moez.QKSMS.ui.MainActivity;
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/data/Conversation.java]: Unused import

File: QKSMS/src/main/java/com/moez/QKSMS/data/Conversation.java [hunk 5]
```diff
@@ -593,6 +594,7 @@ public synchronized boolean hasAttachment() {
      */
     public synchronized boolean hasError() {
         return mHasError;
+
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/data/Conversation.java]: Unnecessary space

File: QKSMS/src/main/java/com/moez/QKSMS/data/ConversationLegacy.java [hunk 6]
```diff
@@ -239,6 +239,7 @@ public int getType() {
         return ids;
     }
 
+
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/data/ConversationLegacy.java]: Unnecessary space

File: QKSMS/src/main/java/com/moez/QKSMS/common/DialogHelper.java (hunk 3 above)
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/common/DialogHelper.java]: Actually this method isn't necessary. There isn't an option to delete failed messages in multi-select mode, so we don't need a method that accepts a list of threadIds. Only the first one with a single threadId is necessary

File: QKSMS/src/main/java/com/moez/QKSMS/transaction/SmsHelper.java [hunk 7]
```diff
@@ -542,6 +542,33 @@ public static int getUnreadMessageCount(Context context) {
         return messages;
     }
 
+    public static List<Message> deleteFailedMessages(Context context) {
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/transaction/SmsHelper.java]: This deletes all failed messages, not the ones for a particular thread. If the dialog that allows you to do this comes from long-pressing on a conversation, then it should only delete the messages from that conversation

File: QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java [hunk 8]
```diff
@@ -195,6 +196,8 @@ public void onItemLongClick(final Conversation conversation, View view) {
 
         dialog.addMenuItem(R.string.menu_delete_conversation, MENU_DELETE_CONVERSATION);
 
+        dialog.addMenuItem(R.string.delete_failed, MENU_DELETE_FAILED);
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java]: Don't need to add this option if the conversation doesn't even have any failed messages

File: QKSMS/src/main/res/values/strings.xml [hunk 9]
```diff
@@ -333,8 +333,12 @@
     <string name="restart_message">These changes will take effect the next time QKSMS is opened</string>
     <!--Dialog title for asking if a user wants to delete a conversation-->
     <string name="delete_conversation">Delete conversation</string>
+    <!--Dialog title for asking if a user wants to delete all failed messages-->
+    <string name="delete_all_failed">Delete conversation</string>
```
Comment (moezbhatti) [inline QKSMS/src/main/res/values/strings.xml]: The string says "Delete conversation" but it should specify that it only deletes failed messages

File: build.gradle [hunk 10]
```diff
@@ -3,6 +3,6 @@ buildscript {
         mavenCentral()
     }
     dependencies {
-        classpath 'com.android.tools.build:gradle:1.0.1'
+        classpath 'com.android.tools.build:gradle:1.2.3'
```
Comment (moezbhatti) [inline build.gradle]: Should actually upgrade to 1.3.0

**Comment by moezbhatti** [comment]
If you want the feature to delete all failed messages, then the option should exist in the overflow menu. When it shows up in the long-press menu of a conversation, it implies that it's contextual and it's only deleting the messages from that conversation

**Comment by CyanBlob** [comment]
I fixed the formatting issues, and it now only deletes the failed messages of the selected conversation.

**Comment by CyanBlob** [comment]
I'm not totally sure how to prevent the dialog from showing if there are no failed messages. I'll work on it.

**Comment by moezbhatti** [comment]
You can replace that line with the following

```
if (conversation.hasError()) {
    dialog.addMenuItem(R.string.delete_failed, MENU_DELETE_FAILED);
}
```

**Comment by moezbhatti** [comment]
That way, the option only shows up if there's a failed message

File: QKSMS/src/main/java/com/moez/QKSMS/transaction/SmsHelper.java [hunk 11]
```diff
@@ -542,6 +542,35 @@ public static int getUnreadMessageCount(Context context) {
         return messages;
     }
 
+    public static List<Message> deleteFailedMessages(Context context, long threadID) {
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/transaction/SmsHelper.java]: Very minor thing, but it should be `threadId` :smile:

**Comment by CyanBlob** [comment]
Implemented. Any other suggestions?

File: QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java [hunk 12]
```diff
@@ -195,6 +196,9 @@ public void onItemLongClick(final Conversation conversation, View view) {
 
         dialog.addMenuItem(R.string.menu_delete_conversation, MENU_DELETE_CONVERSATION);
 
+        if(conversation.hasError())
+            dialog.addMenuItem(R.string.delete_failed, MENU_DELETE_FAILED);
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java]: Part of the code style for the project is to always use braces for if statements, and there should be a space after `if`. Just like how I posted

Sorry, initially this is gonna be kind of annoying but I'm just trying to keep things consistent haha

File: QKSMS/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -346,6 +350,7 @@
 
     <!-- Dialog message text explaining that a draft message will not be saved due to invalid recipients -->
     <string name="discard_message_reason">Your message will be discarded because it has no valid recipients.</string>
+    <string name="delete_failed">Delete failed messages</string>
```
Comment (moezbhatti) [inline QKSMS/src/main/res/values/strings.xml]: Duplicate string - same as the one on line 337

**Comment by CyanBlob** [comment]
Hey, it's no problem. I understand the importance of keeping things consistent, it just might be hard to ignore my coding habits :) But I'll pay more attention to it!

File: QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java [hunk 14]
```diff
@@ -195,6 +196,10 @@ public void onItemLongClick(final Conversation conversation, View view) {
 
         dialog.addMenuItem(R.string.menu_delete_conversation, MENU_DELETE_CONVERSATION);
 
+        if (conversation.hasError()) {
+            dialog.addMenuItem(R.string.delete_failed, MENU_DELETE_FAILED);
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java]: Should be `R.string.delete_all_failed`, this one was deleted :smile:

File: QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java [hunk 15]
```diff
@@ -53,6 +53,7 @@
     private final int MENU_MARK_UNREAD = 5;
     private final int MENU_DELETE_CONVERSATION = 6;
     private final int MENU_MULTI_SELECT = 7;
+    private final int MENU_DELETE_FAILED = 8;
```
Comment (moezbhatti) [inline QKSMS/src/main/java/com/moez/QKSMS/ui/conversationlist/ConversationListFragment.java]: Let's change the order of these

```
     private final int MENU_DELETE_FAILED = 6;
     private final int MENU_DELETE_CONVERSATION = 7;
     private final int MENU_MULTI_SELECT = 8;
```

**Comment by moezbhatti** [comment]
Awesome, looks good to me now

Just tested, works perfectly

**Comment by moezbhatti** [comment]
Anything you wanted to add, or should I go ahead with the merge?

**Comment by moezbhatti** [comment]
Oh, one thing actually

This PR is trying to merge into master
It needs to be merged into develop, not master

**Comment by moezbhatti** [comment]
You'll have to close the pull request, and then create a new one trying to merge into `develop`

**Comment by CyanBlob** [comment]
Done.

**Pull request closed without merging**